0000-richer-documents.rst

```diff
@@ -0,0 +1,460 @@
+one simply cannot guess the flags and hope to get the right annotations. This
+suggests that document consumers should supply an ``SDocContext`` explicitly
+and that the entire annotation consumption API of ``SDoc`` would have to take
+such an argument, e.g ``collectAnns :: SDocContext -> SDoc a -> [a]``.
+
+If, on the other hand, we cannot maintain the invariant described above, then we
+necessarily would have to render the documents (including the annotations) and
+collect all the annotation values as we go, in order to be able to return a list
+of them, e.g
+``showSDocAnns :: (a -> SDoc Void) -> SDocContext -> SDoc a -> (String, [a])``.
+
+An alternative to returning just the annotations would be to
+pair them with the location of their textual rendition, e.g::
+
+  data RenderedAnn a = RenderedAnn
+    { raSpan :: RealSrcSpan -- or a more appropriate type
+    , raAnn  :: a
+    } deriving (Functor, Foldable, Traversable)
+
+  showSDocRenderedAnns
+    :: (a -> SDoc Void)
+    -> SDocContext
+    -> SDoc a
+    -> (String, [RenderedAnn a])
+
+This variant is strictly more general than ``showSDocAnns`` (which can
+be written in terms of ``showSDocRenderedAnns`` by just dropping location
+information) and is implementable regardless of whether our invariant
+holds. The simpler, ``collectAnns`` -style API on the other hand would
+only be guaranteed to work if the invariant holds, and this additional API
+would therefore only be available in that case.
+
+Effect and Interactions
+-----------------------
+
+The main point of adding support for annotations as described
+above is to give a chance to tooling authors to easily access
+AST fragments that today are simply pretty-printed as part of
+some error messages, and this is indeed made possible by this
+proposal. GHC's main error message data type is ``ErrMsg``,
+which contains useful metadata and the actual error message
+document(s), of type ``ErrDoc``.::
+
+    type MsgDoc = SDoc
+
+    data ErrDoc = ErrDoc {
+            -- | Primary error msg.
+            errDocImportant     :: [MsgDoc],
+            -- | Context e.g. \"In the second argument of ...\".
+            errDocContext       :: [MsgDoc],
+            -- | Supplementary information, e.g. \"Relevant bindings include ...\".
+            errDocSupplementary :: [MsgDoc]
+            }
+
+Changing the definition of ``MsgDoc`` to
+``type MsgDoc = SDoc ErrAnnotation`` and "fixing all the
+resulting type errors" will make it possible to build error
+messages that contain annotations. Since such an ``MsgDoc``
+*could* contain annotations but doesn't necessarily have to,
+we could start emitting annotations incrementally, completing
+this effort over several patches, as many as we want.
+
+Updating all the error messages should not be very complicated: the famous
+``Couldn't match expected type`` error message is currently emitted by the
+following code, from ``compiler/typecheck/TcBinds.hs``.::
+
+  typeSigErrMsg :: SDoc
```

Comment:
The code you're looking for is in TcErrors. This little chunk of code is used only in `COMPLETE` pragmas.